crypto/params.c

```diff
@@ -0,0 +1,365 @@
+                                       size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_REAL, buf, sizeof(double),
+                                rsize);
+}
+
+int OSSL_PARAM_get_double(const OSSL_PARAM *p, double *val)
+{
+#define CASE(type) \
+    case sizeof(type): \
+        *val = (double)(*(type *)p->buffer); \
+        return 1
+
+    if (val == NULL || p == NULL)
+        return 0;
+
+    switch (p->data_type) {
+    case OSSL_PARAM_REAL:
+        switch (p->buffer_size) {
+        CASE(double);
+        }
+        break;
+    case OSSL_PARAM_INTEGER:
+        switch (p->buffer_size) {
+        CASE(int32_t);
+        CASE(int64_t);
+        }
+        break;
+    case OSSL_PARAM_UNSIGNED_INTEGER:
+        switch (p->buffer_size) {
+        CASE(uint32_t);
+        CASE(uint64_t);
+        }
+    }
+    return 0;
+#undef CASE
+}
+
+int OSSL_PARAM_set_double(const OSSL_PARAM *p, double val)
+{
+#define CASE(type) \
+    case sizeof(type): \
+        *(type *)p->buffer = (type)val; \
+        SET_RETURN_SIZE(p, sizeof(type)); \
+        return 1
+
+    if (p == NULL)
+        return 0;
+
+    SET_RETURN_SIZE(p, sizeof(double));
+    switch (p->data_type) {
+    case OSSL_PARAM_REAL:
+        switch (p->buffer_size) {
+        CASE(double);
+        }
+        break;
+    case OSSL_PARAM_INTEGER:
+        switch (p->buffer_size) {
+        CASE(int32_t);
+        CASE(int64_t);
+        }
+        break;
+    case OSSL_PARAM_UNSIGNED_INTEGER:
+        switch (p->buffer_size) {
+        CASE(uint32_t);
+        CASE(uint64_t);
+        }
+        break;
+    }
+    return 0;    
+#undef CASE
+}
+
+OSSL_PARAM OSSL_PARAM_construct_utf8_string(const char *key, char *buf,
+                                            size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_UTF8_STRING, buf, bsize,
+                                rsize);
+}
+
+OSSL_PARAM OSSL_PARAM_construct_utf8_string_ptr(const char *key, char **buf,
+                                                size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_UTF8_STRING_PTR, buf, bsize,
+                                rsize);
+}
+
+int OSSL_PARAM_get_utf8_string(const OSSL_PARAM *p, char **val, size_t max_len)
+{
+    char *q;
+    int is_ptr;
+
+    if (val == NULL || p == NULL)
+        return 0;
+    is_ptr = p->data_type != OSSL_PARAM_UTF8_STRING_PTR;
+    if (!is_ptr && p->data_type != OSSL_PARAM_UTF8_STRING)
+        return 0;
+
+    q = *val;
+    if (q == NULL) {
+        if (is_ptr) {
+            *val = *(char **)p->buffer;
+            return 1;
+        }
+        if ((q = OPENSSL_malloc(p->buffer_size)) == NULL)
+            return 0;
+        *val = q;
+        memcpy(q, p->buffer, p->buffer_size);
+        return 1;
+    }
+    if (max_len < p->buffer_size)
+        return 0;
+    if (is_ptr)
+        memcpy(q, *(char **)p->buffer, p->buffer_size);
```

Comment:
Why not `*val = *(char **)p->buffer`?

Reply:
We'd agreed that a copy here was better than replacing the user supplied pointer.
This is the branch where `*val != NULL`.

Reply:
I'm sorry, I don't understand what the difference is between copying and direct assignment here

Reply:
If a caller allocates a buffer and passes it in:
```
char buffer[100];
char *buffer_ptr = buffer;

OSSL_PARAM_get_utf8_string(param, &buffer_ptr, sizeof(buffer));
```
Should this replace the value of `buffer_ptr` with a pointer to the param data?

It should if it was:
```
char *buffer_ptr = NULL;

OSSL_PARAM_get_utf8_string(param, &buffer_ptr, 0);
```

The difference being that the user has allocated space or not.

Reply:
You're missing the point.  At this line, your memcpy *is exactly equivalent to direct assignment*

Reply:
(at least as far as I understand...  what am I missing?)

Reply:
Doesn't the `*(char **)p->buffer` dereference the pointer to get to the pointer to the real data?

Reply:
Or am I missing something here?

Reply:
It dereferences the pointer that the `OSSL_PARAM` owner placed there. In other words, if the `OSSL_PARAM` did this:

``` C
    char *buf_p = NULL;
    OSSL_PARAM whatever[] = {
        { "foo", OSSL_PARAM_UTF8_STRING_PTR, &buf_p, sizeof(buf_p}, NULL },
        { NULL, 0, NULL, 0, NULL }
   } ;

   WHATEVER_set_params(obj, whatever);
```

And down the call stack, this would be called:

``` C
    char *mydata = &something;
    OSSL_PARAM_get_utf8_string(param, &mydata, 0);
```

Then `mydata` will get an exact copy of `buf_p`, i.e. a NULL in this case.

Or at least, that should be the intention

Reply:
So we do want to replace the user's pointer.  I'm confused now.  This isn't what was said in the email yesterday.

Reply:
I think I've got a solution to this little bit of confusion.

I'll split the utf8 and octet string types into two pairs.  The current string types will be unchanged.  The new types being OSSL_PARAM_UTF8_PTR and OSSL_PARAM_OCTET_PTR.  These will deal in pointers directly.  I'll create different functions to get/set them which will more clearly demark the different -- I'm not comfortable having _dual function_ routines that behave _very_ differently for the two types.

Unified string functions would need to behave more similarly IMO.

Reply:
> So we do want to replace the user's pointer. I'm confused now. This isn't what was said in the email yesterday.

I may have expressed myself more vaguely then...  I've changed my language around these things in the last 24 hours, for better precision.

Reply:
> These will deal in pointers directly.

If this will break `OSSL_PARAM` constness, I'm not at all keen

Reply:
No it doesn't break constness.  The OSSL_PARAM still contains a pointer to a pointer to.

I've just pushed a rough sketch of what I thought up.